tests/unit/s2n_recv_test.c

```diff
@@ -219,21 +221,125 @@ int main(int argc, char **argv)
-            EXPECT_EQUAL(s2n_send(client_conn, test_data, sizeof(test_data), &blocked), sizeof(test_data));
+            for (size_t i = 0; i < 3; i++) {
+                EXPECT_EQUAL(s2n_send(client_conn, test_data.data, test_data.size, &blocked), test_data.size);
+            }
 
-            /* Disable multi-recod recv, set legavy behvaior */
+            /* Disable multi-record recv, set legacy behavior */
             EXPECT_SUCCESS(s2n_config_set_recv_multi_record(config, false));
 
-            EXPECT_EQUAL(s2n_recv(server_conn, output, recv_size, &blocked), test_data_size);
+            EXPECT_EQUAL(s2n_recv(server_conn, output.data, recv_size, &blocked), test_data_size);
 
             /* Now enable multi record recv */
             EXPECT_SUCCESS(s2n_config_set_recv_multi_record(config, true));
 
             /* So we should be able to read the remaining two records in a single call */
-            EXPECT_EQUAL(s2n_recv(server_conn, output, recv_size, &blocked), recv_size);
+            EXPECT_EQUAL(s2n_recv(server_conn, output.data, recv_size, &blocked), recv_size);
+        }
+    }
+
+    /* recv blocked status
+     *
+     * This test preserves the `blocked` parameter contract with various states of the connection
+     */
+    {
+        const uint8_t test_data_size = 100;
+        const size_t record_count = 3;
+        DEFER_CLEANUP(struct s2n_blob test_data = { 0 }, s2n_free);
+        EXPECT_SUCCESS(s2n_alloc(&test_data, test_data_size));
+
+        const size_t total_data_size = test_data_size * record_count;
+        DEFER_CLEANUP(struct s2n_blob output = { 0 }, s2n_free);
+        EXPECT_SUCCESS(s2n_alloc(&output, total_data_size));
+
+        s2n_blocked_status blocked = 0;
+
+        DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT),
+                s2n_connection_ptr_free);
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+        DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+                s2n_connection_ptr_free);
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+
+        struct s2n_test_io_pair io_pair = { 0 };
+        EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+        EXPECT_SUCCESS(s2n_connections_set_io_pair(client_conn, server_conn, &io_pair));
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+
+        for (size_t multi_record = 0; multi_record <= 1; multi_record++) {
+            EXPECT_SUCCESS(s2n_config_set_recv_multi_record(config, multi_record));
+            size_t max_recv_size = test_data_size;
+
+            /* in multi-record, we can read all of the records in one go */
+            if (multi_record) {
+                max_recv_size *= record_count;
+            }
+
+            for (size_t read_size = 1; read_size <= total_data_size; read_size++) {
+                /* Write some data across multiple records */
```

Comment:
Nit: Can you not mix comments that start lowercase with comments that start uppercase? It looks kind of strange. I think we pretty much always do uppercase.